nni/compression/pytorch/speedup/compressor.py

```diff
@@ -1,143 +1,336 @@
+        # after manually unpack the tuple/list of tensors, the number of the outputs
+        # of each node should always be one(Except for the TupleUnpack node at the end
+        # of the whole model)
+        assert len(node.outputs) == 1, 'The number of the output should be one after the Tuple unpacked manually'
+
+        out_debugname = node.outputs[0]
+        # update the output mask into self.masks
+        self.masks[out_debugname] = _auto_infer.output_mask
+        self.constant[out_debugname] = _auto_infer.out_constant
+        # update the output result into self.internal_result, so that
+        # the successor nodes can take these output tensors as inputs.
+        self.internal_result[out_debugname] = _auto_infer.output
+        # update the parameter mask of the node
+
+        self.masks[module_name] = _auto_infer.weight_mask
+
+    def update_indirect_sparsity(self, node):
+        """
+        update the indirect sparisty for the target node.
+        """
+        module_name = node.name
+        _logger.info('Update indirect sparsity for %s', module_name)
+        unique_name = node.unique_name
+        if unique_name in self.auto_inferences and self.auto_inferences[unique_name] is not None:
+            # if the auto inference object already in self.auto_inference, then
+            # directly update the previous one
+            # self.auto_inferences[unique_name].update()
+            _logger.info(
+                'Update the indirect sparsity for the %s', unique_name)
+            auto_infer = self.auto_inferences[unique_name]
+            auto_infer.update_indirect_sparsity()
+            # pass the gradient to the predecessor nodes
+            for in_id, tin in enumerate(auto_infer.dummy_input):
+                debug_name = auto_infer.input_debugname[in_id]
+                last_output = self.internal_result[debug_name]
+                # if isinstance(last_output, torch.Tensor):
+                # TODO what if last output is tuple/list of tensor
+                if last_output.grad is not None and tin.grad is not None:
+                    last_output.grad.data += tin.grad.data
+                else:
+                    last_output.grad = tin.grad
+
+    def _vnode_to_value(self, c_node):
+        """
+        translate the C Value node into the values/tensors.
+        """
+        errmsg = "Only support the torch._C.Value type"
+        assert isinstance(c_node, torch._C.Value), errmsg
+        if isinstance(c_node.type(), torch._C.TensorType):
+            shape = tuple(c_node.type().sizes())
+            dtype = c_node.type().scalarType()
+            # TODO should use a more general way to get the input
+            # TODO ugly code here
+            if dtype.startswith('Float') or dtype.startswith('Double'):
+                return torch.rand(shape).to(self.device)
+            else:
+                return torch.randint(0, 10, shape, device=self.device)
```

Comment:
why limit it to 0-10?

Reply:
It may fail on "ReLU6" when this range is too large.

Reply:
add comments to explain why 0-10